examples/sponsored-tx-enabled-auction/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+    let state = State {
+        items:         state_builder.new_map(),
+        cis2_contract: contract,
+        counter:       0,
+    };
+    Ok(state)
+}
+
+/// AddItem entry point to add an item to this contract. To initiate a new
+/// auction, any account can call this entry point. The account initiating the
+/// auction (referred to as the creator) is required to specify the start time,
+/// end time, minimum bid, and the `token_id` associated with the item. At this
+/// stage, the item/auction is assigned the next consecutive index for future
+/// reference.
+#[receive(
+    contract = "sponsored_tx_enabled_auction",
+    name = "addItem",
+    parameter = "AddItemParameter",
+    mutable
+)]
+fn add_item(ctx: &ReceiveContext, host: &mut Host<State>) -> ReceiveResult<()> {
+    // Ensure that only accounts can add an item.
+    let sender_address = match ctx.sender() {
+        Address::Contract(_) => bail!(Error::OnlyAccount.into()),
+        Address::Account(account_address) => account_address,
+    };
+
+    // Getting input parameters.
+    let item: AddItemParameter = ctx.parameter_cursor().get()?;
+
+    // Ensure start < end.
+    ensure!(item.start < item.end, Error::StartEndTimeError.into());
+
+    let slot_time = ctx.metadata().slot_time();
+    // Ensure the auction can run.
+    ensure!(slot_time <= item.end, Error::EndTimeError.into());
+
+    // Assign an index to the item/auction.
+    let counter = host.state_mut().counter;
+    host.state_mut().counter = counter + 1;
+
+    // Insert the item into the state.
+    host.state_mut().items.insert(counter, ItemState {
+        auction_state:  AuctionState::NotSoldYet,
+        highest_bidder: None,
+        name:           item.name,
+        end:            item.end,
+        start:          item.start,
+        highest_bid:    item.minimum_bid,
+        creator:        sender_address,
+        token_id:       item.token_id,
+    });
+
+    Ok(())
+}
+
+/// The `bid` entry point in this contract is not meant to be invoked directly
+/// but rather through the `onCIS2Receive` hook mechanism in the cis2 token
+/// contract. Any account can bid for an item. The `bid` entry point can be
+/// invoked via a sponsored transaction mechanism (`permit` entry point) in case
+/// it is implemented in the cis2 token contract. The bidding flow starts with
+/// an account invoking either the `transfer` or the `permit` entry point in the
+/// cis2 token contract and including the `item_index` in the `additionalData`
+/// section of the input parameter. The `transfer` or the `permit` entry point
+/// will send some token amounts to this contract from the bidder. If the token
+/// amount exceeds the current highest bid, the bid is accepted and the previous
+/// highest bidder is refunded its token investment.
+#[receive(
+    contract = "sponsored_tx_enabled_auction",
+    name = "bid",
+    mutable,
+    parameter = "OnReceivingCis2Params<ContractTokenId, ContractTokenAmount>",
```

Comment:
**Suggestion (non-blocking):** We could have an additional `OnReceivingCis2Params` in `concordium-cis2` with the type of `data` as a generic and provide the `AdditionalDataIndex` for this case, which would result in a schema matching the exact type.

Reply:
Added a new parameter type `OnReceivingCis2DataParams<T, A, D>` in the cis2 library and implemented the `Serial` and `Deserial` trait for it.